// @optimizely-opal/opal-tools-sdk/src/decorators.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import "reflect-metadata";

import { AuthRequirement, Parameter, ParameterType } from "./models";
import { registry } from "./registry";

interface ParameterDefinition {
  description: string;
  name: string;
  required: boolean;
  type: ParameterType;
}

interface ToolOptions {
  authRequirements?: {
    provider: string;
    required?: boolean;
    scopeBundle: string;
  };
  description: string;
  name: string;
  parameters?: ParameterDefinition[];
}

/**
 * Decorator to register a function as an Opal tool
 * @param options Tool options including:
 *   - name: Name of the tool
 *   - description: Description of the tool
 *   - authRequirements: (Optional) Authentication requirements
 *     Format: { provider: "oauth_provider", scopeBundle: "permissions_scope", required: true }
 *     Example: { provider: "google", scopeBundle: "calendar", required: true }
 *
 * Note: If your tool requires authentication, define your handler function with two parameters:
 * ```
 * async function myTool(parameters: ParameterInterface, authData?: any): Promise<any> {
 *   // Your tool implementation
 * }
 * ```
 */
export function tool(options: ToolOptions) {
  return function (
    target: any,
    propertyKey?: string,
    descriptor?: PropertyDescriptor,
  ) {
    const isMethod = propertyKey && descriptor;
    const handler = isMethod ? descriptor.value : target;

    // Generate endpoint from name - ensure hyphens instead of underscores
    const endpoint = `/tools/${options.name.replace(/_/g, "-")}`;

    // Convert parameter definitions to Parameter objects
    const parameters: Parameter[] = [];
    if (options.parameters && options.parameters.length > 0) {
      // Use the explicitly provided parameter definitions
      for (const paramDef of options.parameters) {
        parameters.push(
          new Parameter(
            paramDef.name,
            paramDef.type,
            paramDef.description,
            paramDef.required,
          ),
        );
      }
    }

    // Create auth requirements if specified
    let authRequirements: AuthRequirement[] | undefined;
    if (options.authRequirements) {
      authRequirements = [
        new AuthRequirement(
          options.authRequirements.provider,
          options.authRequirements.scopeBundle,
          options.authRequirements.required ?? true,
        ),
      ];
    }

    // Register the tool with all services
    for (const service of registry.services) {
      service.registerTool(
        options.name,
        options.description,
        handler,
        parameters,
        endpoint,
        authRequirements,
      );
    }

    return isMethod ? descriptor : target;
  };
}
