// @optimizely-opal/opal-tools-sdk/src/models.ts
/**
 * Types of parameters supported by Opal tools
 */
export enum ParameterType {
  Boolean = "boolean",
  Dictionary = "object",
  Integer = "integer",
  List = "array",
  Number = "number",
  String = "string",
}

/**
 * Authentication data for an Opal tool
 */
export type AuthData = {
  credentials: Credentials;
  provider: string;
};

/**
 * Authentication credentials structure
 */
export type Credentials = {
  access_token: string;
  customer_id: string;
  instance_id: string;
  org_sso_id?: string;
  product_sku: string;
};

/**
 * Execution environment for an Opal tool
 */
export type Environment = {
  execution_mode: "headless" | "interactive";
};

/**
 * Authentication requirements for an Opal tool
 */
export class AuthRequirement {
  /**
   * Create a new authentication requirement
   * @param provider Auth provider (e.g., "google", "microsoft")
   * @param scopeBundle Scope bundle (e.g., "calendar", "drive")
   * @param required Whether authentication is required
   */
  constructor(
    public provider: string,
    public scopeBundle: string,
    public required: boolean = true,
  ) {}

  /**
   * Convert to JSON for the discovery endpoint
   */
  toJSON() {
    return {
      provider: this.provider,
      required: this.required,
      scope_bundle: this.scopeBundle,
    };
  }
}

/**
 * Function definition for an Opal tool
 */
export class Function {
  /**
   * HTTP method for the endpoint (default: POST)
   */
  public httpMethod: string = "POST";

  /**
   * Create a new function definition
   * @param name Function name
   * @param description Function description
   * @param parameters Function parameters
   * @param endpoint API endpoint
   * @param authRequirements Authentication requirements (optional)
   */
  constructor(
    public name: string,
    public description: string,
    public parameters: Parameter[],
    public endpoint: string,
    public authRequirements?: AuthRequirement[],
  ) {}

  /**
   * Convert to JSON for the discovery endpoint
   */
  toJSON() {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const result: any = {
      description: this.description,
      endpoint: this.endpoint,
      http_method: this.httpMethod,
      name: this.name,
      parameters: this.parameters.map((p) => p.toJSON()),
    };

    if (this.authRequirements && this.authRequirements.length > 0) {
      result.auth_requirements = this.authRequirements.map((auth) =>
        auth.toJSON(),
      );
    }

    return result;
  }
}

/**
 * Island action definition for interactive UI components
 */
export class IslandAction {
  /**
   * Create a new island action
   * @param name Programmatic action identifier
   * @param label Human-readable button label
   * @param type UI element type
   * @param endpoint API endpoint to call
   * @param operation Operation type
   */
  constructor(
    public name: string,
    public label: string,
    public type: string,
    public endpoint: string,
    public operation: string = "create",
  ) {}

  /**
   * Convert to JSON for the discovery endpoint
   */
  toJSON() {
    return {
      endpoint: this.endpoint,
      label: this.label,
      name: this.name,
      operation: this.operation,
      type: this.type,
    };
  }
}

/**
 * Island field definition for interactive UI components
 */
export class IslandField {
  /**
   * Create a new island field
   * @param name Programmatic field identifier
   * @param label Human-readable label
   * @param type Field type
   * @param value Current field value
   * @param hidden Whether to hide from user
   * @param options Available options for selection
   */
  constructor(
    public name: string,
    public label: string,
    public type: "boolean" | "json" | "string",
    public value: string = "",
    public hidden: boolean = false,
    public options: string[] = [],
  ) {}

  /**
   * Convert to JSON for the discovery endpoint
   */
  toJSON() {
    return {
      hidden: this.hidden,
      label: this.label,
      name: this.name,
      options: this.options,
      type: this.type,
      value: this.value,
    };
  }
}

/**
 * Island configuration for interactive UI components
 */
export class IslandConfig {
  static Action = IslandAction;
  static Field = IslandField;

  /**
   * Create a new island configuration
   * @param fields List of island fields
   * @param actions List of island actions
   * @param type Optional island type
   * @param icon Optional island icon
   */
  constructor(
    public fields: IslandField[],
    public actions: IslandAction[],
    public type?: string,
    public icon?: string,
  ) {}

  /**
   * Convert to JSON for the discovery endpoint
   */
  toJSON() {
    return {
      actions: this.actions.map((action) => action.toJSON()),
      fields: this.fields.map((field) => field.toJSON()),
    };
  }
}

/**
 * Island response configuration
 */
export class IslandResponseConfig {
  /**
   * Create a new island response configuration
   * @param islands List of island configurations
   */
  constructor(public islands: IslandConfig[]) {}

  /**
   * Convert to JSON for the discovery endpoint
   */
  toJSON() {
    return {
      islands: this.islands.map((island) => island.toJSON()),
    };
  }
}

/**
 * Island response wrapper for interactive UI components
 */
export class IslandResponse {
  static ResponseConfig = IslandResponseConfig;

  public type = "island" as const;

  /**
   * Create a new island response
   * @param config Response configuration
   * @param message Optional message for the island response
   */
  constructor(
    public config: IslandResponseConfig,
    public message?: string,
  ) {}

  /**
   * Create an island response with a list of islands
   * @param islands List of island configurations
   * @param message Optional message for the island response
   * @returns New IslandResponse instance
   */
  static create(islands: IslandConfig[], message?: string): IslandResponse {
    return new IslandResponse(new IslandResponseConfig(islands), message);
  }

  /**
   * Convert to JSON for the discovery endpoint
   */
  toJSON() {
    return {
      config: this.config.toJSON(),
    };
  }
}

/**
 * Parameter definition for an Opal tool
 */
export class Parameter {
  /**
   * Create a new parameter definition
   * @param name Parameter name
   * @param type Parameter type
   * @param description Parameter description
   * @param required Whether the parameter is required
   */
  constructor(
    public name: string,
    public type: ParameterType,
    public description: string,
    public required: boolean,
  ) {}

  /**
   * Convert to JSON for the discovery endpoint
   */
  toJSON() {
    return {
      description: this.description,
      name: this.name,
      required: this.required,
      type: this.type,
    };
  }
}
