// @optimizely-opal/opal-tools-sdk/src/service.ts
import express, { Express, Request, Response, Router } from "express";

import { isBlockResponse } from "./block";
import { AuthRequirement, Function, Parameter } from "./models";
import { registry } from "./registry";

export class ToolsService {
  private app: Express;
  private functions: Function[] = [];
  private router: Router;

  /**
   * Initialize a new tools service
   * @param app Express application
   */
  constructor(app: Express) {
    this.app = app;
    this.router = express.Router();
    this.initRoutes();

    // Register this service in the global registry
    registry.services.push(this);
  }

  /**
   * Register a tool function
   * @param name Tool name
   * @param description Tool description
   * @param handler Function implementing the tool
   * @param parameters List of parameters for the tool
   * @param endpoint API endpoint for the tool
   * @param authRequirements Authentication requirements (optional)
   * @param responseType Response type - 'json' (default) or 'block'
   * @param isNewStyle Whether this is a new-style tool (registerTool) vs legacy decorator
   */
  registerTool(
    name: string,
    description: string,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    handler: any, // Changed from Function to any to avoid confusion with built-in Function type
    parameters: Parameter[],
    endpoint: string,
    authRequirements?: AuthRequirement[],
    responseType: "block" | "json" = "json",
    isNewStyle: boolean = false,
  ): void {
    const func = new Function(
      name,
      description,
      parameters,
      endpoint,
      authRequirements,
    );
    this.functions.push(func);

    // Determine if this is a block tool
    const isBlockTool = responseType === "block";

    // Register the actual endpoint
    this.router.post(endpoint, async (req: Request, res: Response) => {
      try {
        console.log(`Received request for ${endpoint}:`, req.body);

        // Extract parameters from the request body
        let params;
        if (req.body && req.body.parameters) {
          // Standard format: { "parameters": { ... } }
          params = req.body.parameters;
          console.log(`Extracted parameters from 'parameters' key:`, params);
        } else {
          // Fallback for direct testing: { "name": "value" }
          console.log(
            `Warning: 'parameters' key not found in request body. Using body directly.`,
          );
          params = req.body;
        }

        // Extract auth data if available
        const authData = req.body && req.body.auth;
        if (authData) {
          console.log(
            `Auth data provided for provider: ${authData.provider || "unknown"}`,
          );
        }

        // Call the handler with extracted parameters
        let result;

        if (isNewStyle) {
          result = await handler(params, {
            mode: (req.body && req.body.execution_mode) || "headless",
            ...(authData && { auth: authData }),
          });
        } else {
          // Check if handler accepts auth as third parameter
          const handlerParamCount = handler.length;

          if (handlerParamCount >= 2) {
            // Handler accepts auth data
            result = await handler(params, authData);
          } else {
            // Handler doesn't accept auth data
            result = await handler(params);
          }
        }

        console.log(`Tool ${name} returned:`, result);

        // Return with appropriate content-type header
        if (isBlockTool) {
          // Validate that block tools return a BlockResponse
          if (!isBlockResponse(result)) {
            throw new Error(
              `Block tool '${name}' must return a BlockResponse object, but returned ${typeof result}`,
            );
          }
          res.set("Content-Type", "application/vnd.opal.block+json");
          res.json(result);
        } else {
          res.json(result);
        }
      } catch (error) {
        console.error(`Error in tool ${name}:`, error);
        res.status(500).json({
          error: error instanceof Error ? error.message : "Unknown error",
        });
      }
    });
  }

  /**
   * Initialize the discovery endpoint
   */
  private initRoutes(): void {
    this.router.get("/discovery", (req: Request, res: Response) => {
      res.json({ functions: this.functions.map((f) => f.toJSON()) });
    });

    this.app.use(this.router);
  }
}
