// @optimizely-opal/opal-tools-sdk/dist/auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.authMiddleware = authMiddleware;
exports.requiresAuth = requiresAuth;
/**
 * Middleware to handle authentication requirements
 * @param req Express request
 * @param res Express response
 * @param next Express next function
 */
function authMiddleware(req, res, next) {
    const authHeader = req.headers.authorization;
    if (!authHeader && req.authRequired) {
        return res.status(401).json({ error: "Authentication required" });
    }
    // The Tools Management Service will provide the appropriate token
    // in the Authorization header
    next();
}
/**
 * Decorator to indicate that a tool requires authentication
 * @param options Authentication options
 */
function requiresAuth(options) {
    return function (target, propertyKey, descriptor) {
        const isMethod = propertyKey && descriptor;
        if (isMethod && descriptor) {
            const originalMethod = descriptor.value;
            descriptor.value = function (...args) {
                // Store auth requirements in function metadata
                const fn = originalMethod;
                fn.__authRequirements__ = {
                    provider: options.provider,
                    required: options.required ?? true,
                    scopeBundle: options.scopeBundle,
                };
                return originalMethod.apply(this, args);
            };
            return descriptor;
        }
        return target;
    };
}
