// @optimizely-opal/opal-tools-sdk/dist/block.js
"use strict";
/**
 * Generated by json-schema-to-typescript
 * DO NOT MODIFY - This file is auto-generated from block-document-spec.json
 * Run 'npm run generate:block' to regenerate
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Block = void 0;
exports.isBlockResponse = isBlockResponse;
/**
 * Builder namespace for Adaptive Block Document components.
 *
 * Usage:
 *   Block.Document({ children: [...] })
 *   Block.Heading({ children: "Title", level: "2" })
 *   Block.Input({ name: "field_name", placeholder: "Enter..." })
 */
exports.Block = {
    Action: (props) => ({
        $type: "Block.Action",
        ...props,
    }),
    Alert: (props) => ({
        $type: "Block.Alert",
        ...props,
    }),
    Badge: (props) => ({
        $type: "Block.Badge",
        ...props,
    }),
    Box: (props) => ({
        $type: "Block.Box",
        ...props,
    }),
    CancelAction: (props) => ({ $type: "Block.CancelAction", ...props }),
    Checkbox: (props) => ({
        $type: "Block.Checkbox",
        ...props,
    }),
    Code: (props) => ({
        $type: "Block.Code",
        ...props,
    }),
    Document: (props) => ({
        $type: "Block.Document",
        ...props,
    }),
    Field: (props) => ({
        $type: "Block.Field",
        ...props,
    }),
    Group: (props) => ({
        $type: "Block.Group",
        ...props,
    }),
    Heading: (props) => ({
        $type: "Block.Heading",
        ...props,
    }),
    Input: (props) => ({
        $type: "Block.Input",
        ...props,
    }),
    Link: (props) => ({
        $type: "Block.Link",
        ...props,
    }),
    Range: (props) => ({
        $type: "Block.Range",
        ...props,
    }),
    Select: (props) => ({
        $type: "Block.Select",
        ...props,
    }),
    Separator: (props) => ({
        $type: "Block.Separator",
        ...props,
    }),
    Switch: (props) => ({
        $type: "Block.Switch",
        ...props,
    }),
    Text: (props) => ({
        $type: "Block.Text",
        ...props,
    }),
    Textarea: (props) => ({
        $type: "Block.Textarea",
        ...props,
    }),
};
/**
 * Type guard to check if a value is a BlockResponse.
 */
function isBlockResponse(value) {
    return (typeof value === "object" &&
        value !== null &&
        ("content" in value ||
            "data" in value ||
            "artifact" in value ||
            "rollback" in value ||
            "error" in value));
}
