// @optimizely-opal/opal-tools-sdk/dist/models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Parameter = exports.IslandResponse = exports.IslandResponseConfig = exports.IslandConfig = exports.IslandField = exports.IslandAction = exports.Function = exports.AuthRequirement = exports.ParameterType = void 0;
/**
 * Types of parameters supported by Opal tools
 */
var ParameterType;
(function (ParameterType) {
    ParameterType["Boolean"] = "boolean";
    ParameterType["Dictionary"] = "object";
    ParameterType["Integer"] = "integer";
    ParameterType["List"] = "array";
    ParameterType["Number"] = "number";
    ParameterType["String"] = "string";
})(ParameterType || (exports.ParameterType = ParameterType = {}));
/**
 * Authentication requirements for an Opal tool
 */
class AuthRequirement {
    /**
     * Create a new authentication requirement
     * @param provider Auth provider (e.g., "google", "microsoft")
     * @param scopeBundle Scope bundle (e.g., "calendar", "drive")
     * @param required Whether authentication is required
     */
    constructor(provider, scopeBundle, required = true) {
        this.provider = provider;
        this.scopeBundle = scopeBundle;
        this.required = required;
    }
    /**
     * Convert to JSON for the discovery endpoint
     */
    toJSON() {
        return {
            provider: this.provider,
            required: this.required,
            scope_bundle: this.scopeBundle,
        };
    }
}
exports.AuthRequirement = AuthRequirement;
/**
 * Function definition for an Opal tool
 */
class Function {
    /**
     * Create a new function definition
     * @param name Function name
     * @param description Function description
     * @param parameters Function parameters
     * @param endpoint API endpoint
     * @param authRequirements Authentication requirements (optional)
     */
    constructor(name, description, parameters, endpoint, authRequirements) {
        this.name = name;
        this.description = description;
        this.parameters = parameters;
        this.endpoint = endpoint;
        this.authRequirements = authRequirements;
        /**
         * HTTP method for the endpoint (default: POST)
         */
        this.httpMethod = "POST";
    }
    /**
     * Convert to JSON for the discovery endpoint
     */
    toJSON() {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const result = {
            description: this.description,
            endpoint: this.endpoint,
            http_method: this.httpMethod,
            name: this.name,
            parameters: this.parameters.map((p) => p.toJSON()),
        };
        if (this.authRequirements && this.authRequirements.length > 0) {
            result.auth_requirements = this.authRequirements.map((auth) => auth.toJSON());
        }
        return result;
    }
}
exports.Function = Function;
/**
 * Island action definition for interactive UI components
 */
class IslandAction {
    /**
     * Create a new island action
     * @param name Programmatic action identifier
     * @param label Human-readable button label
     * @param type UI element type
     * @param endpoint API endpoint to call
     * @param operation Operation type
     */
    constructor(name, label, type, endpoint, operation = "create") {
        this.name = name;
        this.label = label;
        this.type = type;
        this.endpoint = endpoint;
        this.operation = operation;
    }
    /**
     * Convert to JSON for the discovery endpoint
     */
    toJSON() {
        return {
            endpoint: this.endpoint,
            label: this.label,
            name: this.name,
            operation: this.operation,
            type: this.type,
        };
    }
}
exports.IslandAction = IslandAction;
/**
 * Island field definition for interactive UI components
 */
class IslandField {
    /**
     * Create a new island field
     * @param name Programmatic field identifier
     * @param label Human-readable label
     * @param type Field type
     * @param value Current field value
     * @param hidden Whether to hide from user
     * @param options Available options for selection
     */
    constructor(name, label, type, value = "", hidden = false, options = []) {
        this.name = name;
        this.label = label;
        this.type = type;
        this.value = value;
        this.hidden = hidden;
        this.options = options;
    }
    /**
     * Convert to JSON for the discovery endpoint
     */
    toJSON() {
        return {
            hidden: this.hidden,
            label: this.label,
            name: this.name,
            options: this.options,
            type: this.type,
            value: this.value,
        };
    }
}
exports.IslandField = IslandField;
/**
 * Island configuration for interactive UI components
 */
class IslandConfig {
    /**
     * Create a new island configuration
     * @param fields List of island fields
     * @param actions List of island actions
     * @param type Optional island type
     * @param icon Optional island icon
     */
    constructor(fields, actions, type, icon) {
        this.fields = fields;
        this.actions = actions;
        this.type = type;
        this.icon = icon;
    }
    /**
     * Convert to JSON for the discovery endpoint
     */
    toJSON() {
        return {
            actions: this.actions.map((action) => action.toJSON()),
            fields: this.fields.map((field) => field.toJSON()),
        };
    }
}
exports.IslandConfig = IslandConfig;
IslandConfig.Action = IslandAction;
IslandConfig.Field = IslandField;
/**
 * Island response configuration
 */
class IslandResponseConfig {
    /**
     * Create a new island response configuration
     * @param islands List of island configurations
     */
    constructor(islands) {
        this.islands = islands;
    }
    /**
     * Convert to JSON for the discovery endpoint
     */
    toJSON() {
        return {
            islands: this.islands.map((island) => island.toJSON()),
        };
    }
}
exports.IslandResponseConfig = IslandResponseConfig;
/**
 * Island response wrapper for interactive UI components
 */
class IslandResponse {
    /**
     * Create a new island response
     * @param config Response configuration
     * @param message Optional message for the island response
     */
    constructor(config, message) {
        this.config = config;
        this.message = message;
        this.type = "island";
    }
    /**
     * Create an island response with a list of islands
     * @param islands List of island configurations
     * @param message Optional message for the island response
     * @returns New IslandResponse instance
     */
    static create(islands, message) {
        return new IslandResponse(new IslandResponseConfig(islands), message);
    }
    /**
     * Convert to JSON for the discovery endpoint
     */
    toJSON() {
        return {
            config: this.config.toJSON(),
        };
    }
}
exports.IslandResponse = IslandResponse;
IslandResponse.ResponseConfig = IslandResponseConfig;
/**
 * Parameter definition for an Opal tool
 */
class Parameter {
    /**
     * Create a new parameter definition
     * @param name Parameter name
     * @param type Parameter type
     * @param description Parameter description
     * @param required Whether the parameter is required
     */
    constructor(name, type, description, required) {
        this.name = name;
        this.type = type;
        this.description = description;
        this.required = required;
    }
    /**
     * Convert to JSON for the discovery endpoint
     */
    toJSON() {
        return {
            description: this.description,
            name: this.name,
            required: this.required,
            type: this.type,
        };
    }
}
exports.Parameter = Parameter;
